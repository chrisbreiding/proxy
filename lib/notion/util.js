const rp = require('request-promise')

const makeRequest = ({ notionToken, body, method = 'get', path }) => {
  return rp({
    method,
    body,
    uri: `https://api.notion.com/v1/${path}`,
    headers: {
      'Authorization': `Bearer ${notionToken}`,
      'Notion-Version': '2021-05-13',
    },
    json: true,
  })
}

const getBlockChildren = ({ notionToken, pageId }) => {
  return makeRequest({ notionToken, path: `blocks/${pageId}/children` })
}

const appendBlockChildren = ({ notionToken, pageId, blocks }) => {
  return makeRequest({
    notionToken,
    method: 'patch',
    path: `blocks/${pageId}/children`,
    body: {
      children: blocks,
    },
  })
}

const updateBlock = ({ notionToken, blockId, block }) => {
  return makeRequest({
    notionToken,
    method: 'patch',
    path: `blocks/${blockId}`,
    body: block,
  })
}

const getPlainText = (withText) => {
  if (!withText || !withText.text) return

  return withText.text
  .map(({ plain_text }) => plain_text)
  .join('')
  .trim()
}

const weatherIconMap = {
  'clear-day': '☀️',
  'clear-night': '☀️',
  'rain': '☔️',
  'snow': '❄️',
  'sleet': '🌨',
  'wind': '💨',
  'fog': '🌫',
  'cloudy': '☁️',
  'partly-cloudy-day': '⛅️',
  'partly-cloudy-night': '⛅️',
  'default': '🌑',
}

const getWeatherIcon = (iconName) => {
  return weatherIconMap[iconName] || weatherIconMap.default
}

const makeTextPart = (content, color) => {
  const textPart = { text: { content } }

  if (color) {
    textPart.annotations = { color }
  }

  return textPart
}

module.exports = {
  getBlockChildren,
  appendBlockChildren,
  updateBlock,
  getPlainText,
  getWeatherIcon,
  makeTextPart,
}
