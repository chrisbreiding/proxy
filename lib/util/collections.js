const compact = (array) => {
  return array.reduce((memo, item) => {
    return item ? memo.concat(item) : memo
  }, [])
}

module.exports = {
  compact,
}
