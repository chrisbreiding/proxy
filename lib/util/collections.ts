export function clone<T> (obj: T): T {
  return JSON.parse(JSON.stringify(obj))
}

export function compact<T> (array: (T | undefined)[]): T[] {
  return array.reduce((memo: T[], item: T | undefined) => {
    return item ? memo.concat(item) : memo
  }, [] as T[])
}

export async function mapPromisesSerially<T, U> (array: U[], callback: (arg: U) => Promise<T>) {
  const result: T[] = []

  for (const item of array) {
    result.push(await callback(item))
  }

  return result
}

export function times (num: number, filler?: any) {
  return (new Array(num)).fill(filler)
}
