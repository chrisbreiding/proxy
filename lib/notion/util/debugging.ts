import { getBlockPlainText, getRichText } from './general'
import { debug } from '../../util/debug'
import type { NotionBlock } from '../types'

const getDisplayPrefix = (block: NotionBlock) => {
  switch (block.type) {
    case 'paragraph': return ''
    case 'heading_1': return '# '
    case 'heading_2': return '## '
    case 'heading_3': return '### '
    case 'to_do': return `[${'checked' in block.content && block.content.checked ? '✓' : ' '}] `
    case 'bulleted_list_item': return '• '
    case 'numbered_list_item': return '1. '
    case 'toggle': return '> '
    case 'code': return '<> '
    case 'quote': return '| '
    case 'callout': return '" '
    default: return ''
  }
}

function getDisplayText (block: NotionBlock) {
  if (block.type === 'child_page') {
    return `[${'title' in block.content ? block.content.title : '<untitled>'}]`
  }

  if (block.type === 'divider') {
    return '-----'
  }

  if (getRichText(block)) {
    return `${getDisplayPrefix(block)}${getBlockPlainText(block)}`
  }

  return `(${block.type})`
}

// displays block for logging/debugging purposes
export function displayBlocks (blocks: NotionBlock[]) {
  blocks.forEach((block) => {
    debug(`[${block.id}] ${getDisplayText(block)}${block.has_children ? ' (parent)' : ''}`)
  })
}
