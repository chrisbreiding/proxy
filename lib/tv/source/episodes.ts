import dayjs from 'dayjs'

import { makeRequest } from './util'
import { debug, debugVerbose } from '../../util/debug'

export interface TheTVDBEpisode {
  aired: string // "2009-02-17"
  finaleType: string | null
  id: number
  image: string
  imageType: number
  isMovie: number
  lastUpdated: string
  name: string
  nameTranslations: string[]
  number: number
  overview: string
  overviewTranslations: string[]
  runtime: number
  seasonName: string
  seasonNumber: number
  seasons: number | null
  seriesId: number
  year: string
}

export interface Episode {
  airdate?: string
  number: number
  id: string
  season: number
  title: string
}

function convertEpisode (episode: TheTVDBEpisode): Episode {
  return {
    airdate: episode.aired ? dayjs(episode.aired).toISOString() : undefined,
    number: episode.number || 0,
    id: `${episode.id}`,
    season: episode.seasonNumber || 0,
    title: episode.name,
  }
}

interface EpisodesResponse {
  status: string
  data: {
    series: object
    episodes: TheTVDBEpisode[]
  }
  links: {
    prev: null
    self: string
    next: null
    total_items: number
    page_size: number
  }
  token: string
}

async function getAllEpisodesForShow (
  showId: string,
  existingToken?: string,
  page = 0,
  prevEpisodes: TheTVDBEpisode[] = [],
): Promise<TheTVDBEpisode[]> {
  debugVerbose('get episodes for show id %s, page id: %i, total previous: %i', showId, page, prevEpisodes.length)

  const { data, links, token } = await makeRequest({
    path: `series/${showId}/episodes/default`,
    params: { page: `${page}` },
    token: existingToken,
  }) as EpisodesResponse
  const episodes = prevEpisodes.concat(data.episodes)

  if (page === links.next || links.next == null) {
    return episodes
  }

  return getAllEpisodesForShow(showId, token, links.next, episodes)
}

export async function getEpisodesForShow (showId: string): Promise<Episode[]> {
  debugVerbose('get episodes for show id', showId)

  try {
    const episodes = await getAllEpisodesForShow(showId)

    return episodes.map(convertEpisode)
  } catch (error: any) {
    /* c8 ignore next */
    debug(`Getting episodes for show id ${showId} failed:`, error?.stack || error)

    throw error
  }
}

interface TheTvDbEpisodeUpdate {
  entityType: string // "series"
  extraInfo: string
  method: 'create' | 'update' | 'delete'
  methodInt: 1 | 2 | 3 // 1: created, 2: updated, 3: deleted
  recordId: number // 83498
  recordType: string
  seriesId: number
  timeStamp: number // 1667707258
  userId: number
}

export interface EpisodeUpdate {
  id: string
  method: TheTvDbEpisodeUpdate['method']
}

interface ShowAndEpisodeUpdates {
  [key: string]: EpisodeUpdate[]
}

function convertUpdates (updates: ShowAndEpisodeUpdates, episodeUpdate: TheTvDbEpisodeUpdate): ShowAndEpisodeUpdates {
  const showId = `${episodeUpdate.seriesId}`
  const show = updates[showId] || []

  updates[showId] = [
    ...show,
    {
      id: `${episodeUpdate.recordId}`,
      method: episodeUpdate.method,
    },
  ]

  return updates
}

export async function getEpisodesUpdatedSince (date: string): Promise<ShowAndEpisodeUpdates> {
  debugVerbose('find episodes updated since', date)

  try {
    const { data: episodeUpdates } = await makeRequest({
      path: 'updates',
      params: {
        type: 'episodes',
        since: `${dayjs(date).unix()}`,
      },
    })

    return episodeUpdates.reduce(convertUpdates, {} as ShowAndEpisodeUpdates)
  } catch (error: any) {
    debug(`Getting episodes updated since ${date} failed:`, error.stack)

    throw error
  }
}

export async function getEpisode (id: string): Promise<Episode> {
  debugVerbose('Find episode:', id)

  try {
    const { data: episode } = await makeRequest({
      path: `episodes/${id}`,
    }) as { data: TheTVDBEpisode }

    return convertEpisode(episode)
  } catch (error: any) {
    debug(`Getting episode ${id} failed:`, error.stack)

    throw error
  }
}
