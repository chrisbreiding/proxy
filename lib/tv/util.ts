import type { Method } from 'axios'
import { debug, debugVerbose } from '../util/debug'
import { getEnv } from '../util/env'
import { request } from '../util/network'

const apikey = getEnv('THETVDB_API_KEY')
const pin = getEnv('THETVDB_PIN')
export const baseUrl = 'https://api4.thetvdb.com'

interface LoginResult {
  status: string
  data: {
    token: string
  }
}

async function authenticate () {
  debugVerbose('authenticate')

  try {
    const { data, status } = await request({
      method: 'post',
      url: `${baseUrl}/v4/login`,
      body: { apikey, pin },
    }) as LoginResult

    if (!data?.token) {
      throw new Error(`Could not authenticate with TheTVDB, status: ${status}`)
    }

    return data.token as string
  } catch (error: any) {
    debug('Authentication failed:', error?.stack || error)

    throw error
  }
}

interface MakeRequestOptions {
  method?: Method
  params?: { [key: string]: string }
  path: string
  token?: string
}

export async function makeRequest (options: MakeRequestOptions) {
  const { method, params, path } = options
  const token = options.token || await authenticate()
  const url = `${baseUrl}/v4/${path}`

  debugVerbose('request: %o', { method, url, params, token })

  // TODO: handle error
  const result = await request({
    headers: {
      'Authorization': `Bearer ${token}`,
    },
    method,
    params,
    url,
  })

  result.token = token

  return result
}
