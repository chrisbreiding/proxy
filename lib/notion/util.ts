import type { Method } from 'axios'
import dayjs from 'dayjs'
import type {
  ApiColor,
  BlockObjectResponse,
  BookmarkBlockObjectResponse,
  BulletedListItemBlockObjectResponse,
  CalloutBlockObjectResponse,
  ChildDatabaseBlockObjectResponse,
  ChildPageBlockObjectResponse,
  ColumnBlockObjectResponse,
  ColumnListBlockObjectResponse,
  DividerBlockObjectResponse,
  EmbedBlockObjectResponse,
  EquationBlockObjectResponse,
  FileBlockObjectResponse,
  Heading1BlockObjectResponse,
  Heading2BlockObjectResponse,
  Heading3BlockObjectResponse,
  ImageBlockObjectResponse,
  LinkPreviewBlockObjectResponse,
  LinkToPageBlockObjectResponse,
  ListBlockChildrenResponse,
  ListDatabasesResponse,
  NumberedListItemBlockObjectResponse,
  ParagraphBlockObjectResponse,
  PdfBlockObjectResponse,
  QuoteBlockObjectResponse,
  RichTextItemResponse,
  SyncedBlockBlockObjectResponse,
  TableBlockObjectResponse,
  TableOfContentsBlockObjectResponse,
  TableRowBlockObjectResponse,
  TemplateBlockObjectResponse,
  ToDoBlockObjectResponse,
  ToggleBlockObjectResponse,
  UnsupportedBlockObjectResponse,
  UpdateBlockBodyParameters,
  VideoBlockObjectResponse,
} from '@notionhq/client/build/src/api-endpoints'

import { clone, compact } from '../util/collections'
import { debug } from '../util/debug'
import { request } from '../util/network'

export const dateRegex = /(?:Sun|Mon|Tue|Wed|Thu|Fri|Sat), (\d{1,2}\/\d{1,2})/

export function getDateFromText (dateText: string) {
  const currentDate = dayjs()
  // originally assume the date's year matches the current year
  const date = dayjs(`${dateText}/${currentDate.year()}`, 'M/D/YYYY')
  // if the current month is after the date's month, we've crossed over years
  // and the date's year should be the next year. usually, it's because it's
  // currently December, but the date's month is January. this can be
  // generically applied, so it will work if it's currently June, but we're
  // trying to get the weather for April. we don't want past weather, only
  // future weather, so assume it's for the following year.
  if (currentDate.month() > date.month()) {
    return date.add(1, 'year')
  }

  return date
}

interface MakeRequestOptions {
  body?: object
  method?: Method
  notionToken: string
  path: string
}

function makeRequest<T> (options: MakeRequestOptions): Promise<T> {
  const { notionToken, body, path, method = 'get' } = options

  return request({
    method,
    body,
    url: `https://api.notion.com/v1/${path}`,
    headers: {
      'Authorization': `Bearer ${notionToken}`,
      'Notion-Version': '2022-06-28',
    },
  })
}

interface GetBlockChildrenOptions {
  notionToken: string
  pageId: string
}

export async function getBlockChildren ({ notionToken, pageId }: GetBlockChildrenOptions) {
  const { results } = await makeRequest<ListBlockChildrenResponse>({ notionToken, path: `blocks/${pageId}/children` }) as { results: BlockObjectResponse[] }

  return results.map(convertNotionBlockToOwnBlock)
}

interface GetBlockChildrenDeepOptions {
  filter?: (block: NotionBlock) => boolean
  includeId?: boolean
  notionToken: string
  pageId: string
}

export async function getBlockChildrenDeep ({ notionToken, pageId, filter, includeId }: GetBlockChildrenDeepOptions): Promise<NotionBlock[]> {
  const blocks = await getBlockChildren({ notionToken, pageId })
  const filteredBlocks = filter ? blocks.filter(filter) : blocks
  const blocksWithChildren = await Promise.all(filteredBlocks.map(async (block) => {
    if (block.has_children && block.type !== 'child_page') {
      block.children = await getBlockChildrenDeep({ notionToken, pageId: block.id, includeId })
    }

    return block
  }))

  return compact(blocksWithChildren)
}

type Content = BookmarkBlockObjectResponse['bookmark']
| BulletedListItemBlockObjectResponse['bulleted_list_item']
| CalloutBlockObjectResponse['callout']
| ChildDatabaseBlockObjectResponse['child_database']
| ChildPageBlockObjectResponse['child_page']
| ColumnBlockObjectResponse['column']
| ColumnListBlockObjectResponse['column_list']
| DividerBlockObjectResponse['divider']
| EmbedBlockObjectResponse['embed']
| EquationBlockObjectResponse['equation']
| FileBlockObjectResponse['file']
| Heading1BlockObjectResponse['heading_1']
| Heading2BlockObjectResponse['heading_2']
| Heading3BlockObjectResponse['heading_3']
| ImageBlockObjectResponse['image']
| LinkPreviewBlockObjectResponse['link_preview']
| LinkToPageBlockObjectResponse['link_to_page']
| NumberedListItemBlockObjectResponse['numbered_list_item']
| ParagraphBlockObjectResponse['paragraph']
| PdfBlockObjectResponse['pdf']
| QuoteBlockObjectResponse['quote']
| SyncedBlockBlockObjectResponse['synced_block']
| TableBlockObjectResponse['table']
| TableOfContentsBlockObjectResponse['table_of_contents']
| TableRowBlockObjectResponse['table_row']
| TemplateBlockObjectResponse['template']
| ToDoBlockObjectResponse['to_do']
| ToggleBlockObjectResponse['toggle']
| UnsupportedBlockObjectResponse['unsupported']
| VideoBlockObjectResponse['video']

export interface OwnBlock {
  children?: OwnBlock[] | NotionBlock[]
  type: BlockObjectResponse['type']
  content: Content
}

export interface NotionBlock extends OwnBlock {
  children?: NotionBlock[]
  has_children: boolean
  id: string
}

export type Block = OwnBlock | NotionBlock

function getRichText (block: OwnBlock): RichTextItemResponse[] | undefined {
  return 'rich_text' in block.content ? block.content.rich_text : undefined
}

interface MakeAppendRequestOptions {
  blocks: OutgoingBlock[]
  notionToken: string
  pageId: string
}

export function makeAppendRequest ({ notionToken, pageId, blocks }: MakeAppendRequestOptions) {
  return makeRequest<ListBlockChildrenResponse>({
    notionToken,
    method: 'patch',
    path: `blocks/${pageId}/children`,
    body: {
      children: blocks,
    },
  })
  // don't understand why this fails converage
  /* c8 ignore next */
}

interface AppendBlockChildrenOptions {
  blocks: OwnBlock[]
  notionToken: string
  pageId: string
}

export async function appendBlockChildrenDeep ({ notionToken, pageId, blocks }: AppendBlockChildrenOptions) {
  function moveChildren (blocks: NotionBlock[] | OwnBlock[]) {
    return blocks.map((block) => {
      const convertedBlock = convertBlockToOutgoingBlock(block)

      if (!block.children) return convertedBlock

      // @ts-ignore
      convertedBlock[block.type].children = moveChildren(block.children)

      return convertedBlock
    })
  }

  return makeAppendRequest({ notionToken, pageId, blocks: moveChildren(blocks) })
}

export async function appendBlockChildren ({ notionToken, pageId, blocks }: AppendBlockChildrenOptions) {
  let toAppend: OutgoingBlock[] = []

  for (const block of blocks) {
    const convertedBlock = convertBlockToOutgoingBlock(block)

    if (block.children) {
      const children = block.children

      toAppend.push(convertedBlock)

      const { results } = await makeAppendRequest({ notionToken, pageId, blocks: toAppend })
      toAppend = []

      await appendBlockChildren({
        notionToken,
        pageId: results[results.length - 1].id,
        blocks: children,
      })
    } else {
      toAppend.push(convertedBlock)
    }
  }

  if (toAppend.length) {
    return makeAppendRequest({
      notionToken,
      pageId,
      blocks: toAppend,
    })
  }
}

interface UpdateBlockOptions {
  notionToken: string
  blockId: string
  block: Block
}

export function updateBlock ({ notionToken, blockId, block }: UpdateBlockOptions) {
  const convertedBlock = convertBlockToOutgoingBlock(block)

  return makeRequest<void>({
    notionToken,
    method: 'patch',
    path: `blocks/${blockId}`,
    body: convertedBlock,
  })
}

interface UpdatePageOptions {
  notionToken: string
  pageId: string
  properties: {
    [key: string]: {
      date: {
        start: string
      }
    }
  }
}

export function updatePage ({ notionToken, pageId, properties }: UpdatePageOptions) {
  return makeRequest<void>({
    notionToken,
    body: { properties },
    method: 'patch',
    path: `pages/${pageId}`,
  })
}

interface QueryDatabasesOptions {
  notionToken: string
  databaseId: string
}

export function queryDatabases ({ notionToken, databaseId }: QueryDatabasesOptions) {
  return makeRequest<ListDatabasesResponse>({
    notionToken,
    method: 'post',
    path: `databases/${databaseId}/query`,
  })
}

export function convertNotionBlockToOwnBlock (block: BlockObjectResponse): NotionBlock {
  const type = block.type

  return {
    // @ts-ignore
    content: clone(block[type] as Content),
    has_children: block.has_children,
    id: block.id,
    type,
  }
  // don't understand why this fails converage
  /* c8 ignore next */
}

type OutgoingBlock = UpdateBlockBodyParameters & {
  object: string
  type: BlockObjectResponse['type']
}

export function convertBlockToOutgoingBlock (block: Block): OutgoingBlock {
  return {
    object: 'block',
    type: block.type,
    [block.type]: block.content,
  }
  // don't understand why this fails converage
  /* c8 ignore next */
}

interface MakeBlockOptions {
  annotations?: Partial<RichTextItemResponse['annotations']>
  children?: OwnBlock[] | NotionBlock[]
  text: string
  type: BlockObjectResponse['type']
}

export function makeBlock ({ text, type = 'paragraph', annotations, children }: MakeBlockOptions) {
  return {
    type,
    children,
    content: {
      rich_text: [
        {
          type: 'text',
          text: {
            content: text,
          },
          annotations,
          plain_text: text,
        },
      ],
    },
  } as OwnBlock
}

export function richTextToPlainText (richText: RichTextItemResponse[]) {
  return richText
  .map(({ plain_text }) => plain_text)
  .join('')
  .trim()
}

export function getBlockPlainText (block: Block) {
  const richText = getRichText(block)

  if (!richText) return

  return richTextToPlainText(richText)
}

interface RichTextItem {
  text: {
    content: string
    link?: { url: string } | null
  }
  type?: 'text'
  annotations?: {
    bold?: boolean
    italic?: boolean
    strikethrough?: boolean
    underline?: boolean
    code?: boolean
    color?: ApiColor
  }
}

export function makeTextPart (content: string, color?: ApiColor) {
  const textPart = { text: { content } } as RichTextItem

  if (color) {
    textPart.annotations = { color }
  }

  return textPart
}

export function getMonthNameFromIndex (monthIndex: number, short = false) {
  return dayjs().month(monthIndex).format(short ? 'MMM' : 'MMMM')
}

export function getMonths ({ short }: { short?: boolean } = {}) {
  return Array.from(new Array(12)).map((_, i) => {
    return getMonthNameFromIndex(i, short)
  })
}

/* c8 ignore start */
const getDisplayPrefix = (block: NotionBlock) => {
  switch (block.type) {
    case 'paragraph': return ''
    case 'heading_1': return '# '
    case 'heading_2': return '## '
    case 'heading_3': return '### '
    case 'to_do': return `[${'checked' in block.content && block.content.checked ? '✓' : ' '}] `
    case 'bulleted_list_item': return '• '
    case 'numbered_list_item': return '1. '
    case 'toggle': return '> '
    case 'code': return '<> '
    case 'quote': return '| '
    case 'callout': return '" '
    default: return ''
  }
}

function getDisplayText (block: NotionBlock) {
  if (block.type === 'child_page') {
    return `[${'title' in block.content ? block.content.title : '<untitled>'}]`
  }

  if (block.type === 'divider') {
    return '-----'
  }

  if (getRichText(block)) {
    return `${getDisplayPrefix(block)}${getBlockPlainText(block)}`
  }

  return `(${block.type})`
}

// displays block for logging/debugging purposes
export function displayBlocks (blocks: NotionBlock[]) {
  blocks.forEach((block) => {
    debug(`[${block.id}] ${getDisplayText(block)}${block.has_children ? ' (parent)' : ''}`)
  })
}
/* c8 ignore stop */
