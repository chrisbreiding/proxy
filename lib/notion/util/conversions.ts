import type { BlockObjectResponse, RichTextItemResponse } from '@notionhq/client/build/src/api-endpoints'
import type { Block, NotionBlock, OutgoingBlock } from '../types'
import { clone } from '../../util/collections'

export function convertNotionBlockToOwnBlock (block: BlockObjectResponse): NotionBlock {
  const type = block.type
  // @ts-ignore
  const content = block[type] as Content

  return {
    content: clone(content),
    has_children: block.has_children,
    id: block.id,
    type,
  }
}

export function convertBlockToOutgoingBlock (block: Block): OutgoingBlock {
  return {
    object: 'block',
    type: block.type,
    [block.type]: block.content,
  }
}

export function richTextToPlainText (richText: RichTextItemResponse[]) {
  return richText
  .map(({ plain_text }) => plain_text)
  .join('')
  .trim()
}
