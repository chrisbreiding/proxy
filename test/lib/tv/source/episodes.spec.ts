import nock from 'nock'
import { describe, expect, it } from 'vitest'

const apikey = process.env.THETVDB_API_KEY = 'api-key'
const pin = process.env.THETVDB_PIN = 'pin'

import { baseUrl } from '../../../../lib/tv/source/util'
import { fixtureContents } from '../../../support/util'
import { getEpisodesForShow } from '../../../../lib/tv/source/episodes'
import { nockLogin } from '../util'
import { clone } from '../../../../lib/util/collections'

const episodes = fixtureContents('tv/episodes')

describe('lib/tv/source/episodes', () => {
  describe('#getEpisodesForShow', () => {
    it('returns episodes for show', async () => {
      nockLogin(apikey, pin)

      nock(baseUrl)
      .matchHeader('Authorization', 'Bearer token')
      .get('/v4/series/12345/episodes/default?page=0')
      .reply(200, episodes)

      const result = await getEpisodesForShow('12345')

      expect(result).toMatchSnapshot()
    })

    it('returns all episodes for show if there are paginated results', async () => {
      nockLogin(apikey, pin)

      const page0Episodes = clone(episodes)
      page0Episodes.links.next = 1

      nock(baseUrl)
      .matchHeader('Authorization', 'Bearer token')
      .get('/v4/series/67890/episodes/default?page=0')
      .reply(200, page0Episodes)

      const page1Episodes = clone(episodes)
      page1Episodes.links.prev = 0

      nock(baseUrl)
      .matchHeader('Authorization', 'Bearer token')
      .get('/v4/series/67890/episodes/default?page=1')
      .reply(200, page1Episodes)

      const result = await getEpisodesForShow('67890')

      expect(result).toMatchSnapshot()
    })
  })
})
