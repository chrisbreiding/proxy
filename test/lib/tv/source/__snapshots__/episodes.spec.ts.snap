// Vitest Snapshot v1

exports[`lib/tv/source/episodes > #getEpisodesForShow > returns all episodes for show if there are paginated results 1`] = `
[
  {
    "airdate": "2009-02-17T05:00:00.000Z",
    "episodeNumber": 1,
    "id": "3859781",
    "season": 0,
    "title": "Good Cop Bad Cop",
  },
  {
    "airdate": undefined,
    "episodeNumber": 0,
    "id": "3859791",
    "season": 0,
    "title": "Wedding Day",
  },
  {
    "airdate": "2009-02-17T05:00:00.000Z",
    "episodeNumber": 3,
    "id": "3859801",
    "season": 0,
    "title": "TwaughtHammer",
  },
  {
    "airdate": "2009-02-17T05:00:00.000Z",
    "episodeNumber": 4,
    "id": "3859811",
    "season": 0,
    "title": "Marie's Confession",
  },
  {
    "airdate": "2009-02-17T05:00:00.000Z",
    "episodeNumber": 5,
    "id": "3859821",
    "season": 0,
    "title": "The Break-In",
  },
  {
    "airdate": "2009-02-17T05:00:00.000Z",
    "episodeNumber": 1,
    "id": "3859781",
    "season": 0,
    "title": "Good Cop Bad Cop",
  },
  {
    "airdate": undefined,
    "episodeNumber": 0,
    "id": "3859791",
    "season": 0,
    "title": "Wedding Day",
  },
  {
    "airdate": "2009-02-17T05:00:00.000Z",
    "episodeNumber": 3,
    "id": "3859801",
    "season": 0,
    "title": "TwaughtHammer",
  },
  {
    "airdate": "2009-02-17T05:00:00.000Z",
    "episodeNumber": 4,
    "id": "3859811",
    "season": 0,
    "title": "Marie's Confession",
  },
  {
    "airdate": "2009-02-17T05:00:00.000Z",
    "episodeNumber": 5,
    "id": "3859821",
    "season": 0,
    "title": "The Break-In",
  },
]
`;

exports[`lib/tv/source/episodes > #getEpisodesForShow > returns episodes for show 1`] = `
[
  {
    "airdate": "2009-02-17T05:00:00.000Z",
    "episodeNumber": 1,
    "id": "3859781",
    "season": 0,
    "title": "Good Cop Bad Cop",
  },
  {
    "airdate": undefined,
    "episodeNumber": 0,
    "id": "3859791",
    "season": 0,
    "title": "Wedding Day",
  },
  {
    "airdate": "2009-02-17T05:00:00.000Z",
    "episodeNumber": 3,
    "id": "3859801",
    "season": 0,
    "title": "TwaughtHammer",
  },
  {
    "airdate": "2009-02-17T05:00:00.000Z",
    "episodeNumber": 4,
    "id": "3859811",
    "season": 0,
    "title": "Marie's Confession",
  },
  {
    "airdate": "2009-02-17T05:00:00.000Z",
    "episodeNumber": 5,
    "id": "3859821",
    "season": 0,
    "title": "The Break-In",
  },
]
`;
